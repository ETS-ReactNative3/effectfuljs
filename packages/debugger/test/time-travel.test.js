require("./setup-time-travel");
const trace = require("./run").bind(null, true);
const D = require("../main");

describe("time traveling", function() {
  test("Object tracing", function() {
    D.TimeTravel.reset();
    const s = Symbol("S");
    const v = D.wrap({ a: "A", b: "B", c: "C", [s]: "S" });
    const protoUnwrapped = {};
    const proto = D.wrap(protoUnwrapped);
    const orig = Object.setPrototypeOf({ ...v }, proto);
    Object.setPrototypeOf(v, proto);
    expect(Object.getPrototypeOf(v)).toBe(proto);
    D.TimeTravel.checkpoint();
    delete v.a;
    v.a = "a";
    delete v.b;
    v[s] = "s";
    Object.setPrototypeOf(v, null);
    expect(v).toStrictEqual(
      Object.assign(Object.create(null), {
        a: "a",
        c: "C",
        [s]: "s",
        [D.TimeTravel.traceDataSymbol]: v[D.TimeTravel.traceDataSymbol]
      })
    );
    D.TimeTravel.checkpoint();
    D.TimeTravel.undo();
    expect(v).toStrictEqual(orig);
  });
  test("Array tracing", function() {
    D.TimeTravel.reset();
    const arr = D.wrap([1, 2, 3, 4, 5]);
    const orig = [...arr];
    D.TimeTravel.checkpoint();
    arr.splice(2, 2, "A", "B", "C");
    expect([...arr]).toStrictEqual([1, 2, "A", "B", "C", 5]);
    D.TimeTravel.checkpoint();
    D.TimeTravel.undo();
    expect([...arr]).toStrictEqual(orig);
    arr.push("Z");
    arr.unshift("A");
    expect([...arr]).toStrictEqual(["A", 1, 2, 3, 4, 5, "Z"]);
    D.TimeTravel.checkpoint();
    D.TimeTravel.undo();
    expect([...arr]).toEqual(orig);
    arr[0] = "A";
    arr[4] = "Z";
    expect([...arr]).toEqual(["A", 2, 3, 4, "Z"]);
    D.TimeTravel.checkpoint();
    D.TimeTravel.undo();
    expect([...arr]).toEqual(orig);
  });
  test("Map tracing", function() {
    D.TimeTravel.reset();
    const map = D.wrap(new Map([[1, "A"], [2, "B"], [3, "C"]]));
    map.deb_hi = "there";
    const orig = [...map];
    D.TimeTravel.checkpoint();
    map.deb_hi = "THERE";
    map.set(1, "D").delete(2);
    map.delete(3);
    map.set(3, "E").set(4, "F");
    D.TimeTravel.checkpoint();
    D.TimeTravel.undo();
    expect(map.has(4)).toBe(false);
    expect(map.deb_hi).toBe("there");
    expect(map.get(3)).toBe("C");
    D.TimeTravel.checkpoint();
    expect([...map]).toEqual(orig);
    map.clear();
    D.TimeTravel.checkpoint();
    D.TimeTravel.undo();
    expect([...map]).toEqual(orig);
    map.clear();
    map.set(2, "E");
    map.set(1, "F");
    D.TimeTravel.checkpoint();
    D.TimeTravel.undo();
    expect([...map]).toEqual(orig);
  });
  test("Set tracing", function() {
    D.TimeTravel.reset();
    const set = D.wrap(new Set([1, 2, 3]));
    const orig = [...set];
    set.deb_hi = "there";
    D.TimeTravel.checkpoint();
    set.deb_hi = "THERE";
    set.delete(1);
    set.delete(3);
    set.add(4);
    set.add(2);
    set.add(1);
    D.TimeTravel.checkpoint();
    D.TimeTravel.undo();
    expect(set.has(4)).toBe(false);
    expect(set.deb_hi).toBe("there");
    expect(set.has(3)).toBe(true);
    expect([...set]).toEqual(orig);
    set.clear();
    D.TimeTravel.checkpoint();
    D.TimeTravel.undo();
    expect([...set]).toEqual(orig);
    set.clear();
    set.add(8);
    D.TimeTravel.checkpoint();
    D.TimeTravel.undo();
    expect([...set]).toEqual(orig);
  });
  test("captured state", function() {
    D.TimeTravel.reset();
    global.console = { log: jest.fn() };
    D.evalThunk(require("./__fixtures__/counters"));
    const mod = trace();
    mod.getCounter();
    const [counter, tr, dir, set] = trace(true);
    console.log("INIT", tr, dir, set);
    let stepNum = 0;
    function counterStep() {
      counter();
      console.log(
        "STEP",
        stepNum++,
        trace(true),
        JSON.stringify({ tr, dir: [...dir], set: [...set] })
      );
    }
    counterStep();
    counterStep();
    counterStep();
    console.log("SAVE", tr, dir, set);
    D.TimeTravel.checkpoint();
    counterStep();
    counterStep();
    counterStep();
    counterStep();
    console.log("UNDO");
    D.TimeTravel.checkpoint();
    D.TimeTravel.undo();
    counterStep();
    counterStep();
    console.log("SAVE2");
    D.TimeTravel.checkpoint();
    counterStep();
    counterStep();
    console.log("UNDO SAVE2");
    D.TimeTravel.checkpoint();
    D.TimeTravel.undo();
    D.TimeTravel.undo();
    counterStep();
    counterStep();
    console.log("UNDO SAVE");
    counter();
    console.log(trace(true));
    console.log("UNDO SAVE1");
    D.TimeTravel.checkpoint();
    D.TimeTravel.undo();
    counter();
    console.log(trace());
    expect(console.log.mock.calls).toMatchSnapshot();
  });
});
