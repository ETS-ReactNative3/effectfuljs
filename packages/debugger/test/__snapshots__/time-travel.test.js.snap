// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`time traveling captured state 1`] = `
Array [
  Array [
    "INIT",
    Array [
      Object {
        "i": 0,
        "u": 3,
      },
      Object {
        "i": 1,
        "u": 13,
      },
      Object {
        "i": 2,
        "u": 23,
      },
      Object {
        "i": 3,
        "u": 33,
      },
      Object {
        "i": 4,
        "u": 43,
      },
      Object {
        "i": 5,
        "u": 53,
      },
      Object {
        "i": 6,
        "u": 63,
      },
      Object {
        "i": 3,
        "u": 33,
      },
      Object {
        "i": 4,
        "u": 43,
      },
      Object {
        "i": 5,
        "u": 53,
      },
      Object {
        "i": 6,
        "u": 63,
      },
      Object {
        "i": 3,
        "u": 33,
      },
      Object {
        "i": 4,
        "u": 43,
      },
      Object {
        "i": 5,
        "u": 53,
      },
      Object {
        "i": 3,
        "u": 33,
      },
    ],
    Map {
      Object {
        "i": 0,
        "u": 3,
      } => 0,
      Object {
        "i": 1,
        "u": 13,
      } => 1,
      Object {
        "i": 2,
        "u": 23,
      } => 2,
      Object {
        "i": 3,
        "u": 33,
      } => 3,
      Object {
        "i": 4,
        "u": 43,
      } => 4,
      Object {
        "i": 5,
        "u": 53,
      } => 5,
      Object {
        "i": 6,
        "u": 63,
      } => 6,
      Object {
        "i": 3,
        "u": 33,
      } => 3,
      Object {
        "i": 4,
        "u": 43,
      } => 4,
      Object {
        "i": 5,
        "u": 53,
      } => 5,
      Object {
        "i": 6,
        "u": 63,
      } => 6,
      Object {
        "i": 3,
        "u": 33,
      } => 3,
      Object {
        "i": 4,
        "u": 43,
      } => 4,
      Object {
        "i": 5,
        "u": 53,
      } => 5,
      Object {
        "i": 3,
        "u": 33,
      } => 3,
    },
    Set {
      0,
      1,
      2,
      3,
      4,
      5,
      6,
    },
  ],
  Array [
    "STEP",
    0,
    0,
    "{\\"tr\\":[{\\"i\\":0,\\"u\\":0}],\\"dir\\":[[{\\"i\\":0,\\"u\\":0},0]],\\"set\\":[0]}",
  ],
  Array [
    "STEP",
    1,
    1,
    "{\\"tr\\":[{\\"i\\":0,\\"u\\":1},{\\"i\\":1,\\"u\\":11}],\\"dir\\":[[{\\"i\\":0,\\"u\\":1},0],[{\\"i\\":1,\\"u\\":11},1]],\\"set\\":[0,1]}",
  ],
  Array [
    "STEP",
    2,
    2,
    "{\\"tr\\":[{\\"i\\":0,\\"u\\":2},{\\"i\\":1,\\"u\\":12},{\\"i\\":2,\\"u\\":22}],\\"dir\\":[[{\\"i\\":0,\\"u\\":2},0],[{\\"i\\":1,\\"u\\":12},1],[{\\"i\\":2,\\"u\\":22},2]],\\"set\\":[0,1,2]}",
  ],
  Array [
    "SAVE",
    Array [
      Object {
        "i": 0,
        "u": 3,
      },
      Object {
        "i": 1,
        "u": 13,
      },
      Object {
        "i": 2,
        "u": 23,
      },
      Object {
        "i": 3,
        "u": 33,
      },
      Object {
        "i": 4,
        "u": 43,
      },
      Object {
        "i": 5,
        "u": 53,
      },
      Object {
        "i": 6,
        "u": 63,
      },
      Object {
        "i": 3,
        "u": 33,
      },
      Object {
        "i": 4,
        "u": 43,
      },
      Object {
        "i": 5,
        "u": 53,
      },
      Object {
        "i": 6,
        "u": 63,
      },
      Object {
        "i": 3,
        "u": 33,
      },
      Object {
        "i": 4,
        "u": 43,
      },
      Object {
        "i": 5,
        "u": 53,
      },
      Object {
        "i": 3,
        "u": 33,
      },
    ],
    Map {
      Object {
        "i": 0,
        "u": 3,
      } => 0,
      Object {
        "i": 1,
        "u": 13,
      } => 1,
      Object {
        "i": 2,
        "u": 23,
      } => 2,
      Object {
        "i": 3,
        "u": 33,
      } => 3,
      Object {
        "i": 4,
        "u": 43,
      } => 4,
      Object {
        "i": 5,
        "u": 53,
      } => 5,
      Object {
        "i": 6,
        "u": 63,
      } => 6,
      Object {
        "i": 3,
        "u": 33,
      } => 3,
      Object {
        "i": 4,
        "u": 43,
      } => 4,
      Object {
        "i": 5,
        "u": 53,
      } => 5,
      Object {
        "i": 6,
        "u": 63,
      } => 6,
      Object {
        "i": 3,
        "u": 33,
      } => 3,
      Object {
        "i": 4,
        "u": 43,
      } => 4,
      Object {
        "i": 5,
        "u": 53,
      } => 5,
      Object {
        "i": 3,
        "u": 33,
      } => 3,
    },
    Set {
      0,
      1,
      2,
      3,
      4,
      5,
      6,
    },
  ],
  Array [
    "STEP",
    3,
    3,
    "{\\"tr\\":[{\\"i\\":0,\\"u\\":3},{\\"i\\":1,\\"u\\":13},{\\"i\\":2,\\"u\\":23},{\\"i\\":3,\\"u\\":33}],\\"dir\\":[[{\\"i\\":0,\\"u\\":3},0],[{\\"i\\":1,\\"u\\":13},1],[{\\"i\\":2,\\"u\\":23},2],[{\\"i\\":3,\\"u\\":33},3]],\\"set\\":[0,1,2,3]}",
  ],
  Array [
    "STEP",
    4,
    4,
    "{\\"tr\\":[{\\"i\\":0,\\"u\\":4},{\\"i\\":1,\\"u\\":14},{\\"i\\":2,\\"u\\":24},{\\"i\\":3,\\"u\\":34},{\\"i\\":4,\\"u\\":44}],\\"dir\\":[[{\\"i\\":0,\\"u\\":4},0],[{\\"i\\":1,\\"u\\":14},1],[{\\"i\\":2,\\"u\\":24},2],[{\\"i\\":3,\\"u\\":34},3],[{\\"i\\":4,\\"u\\":44},4]],\\"set\\":[0,1,2,3,4]}",
  ],
  Array [
    "STEP",
    5,
    5,
    "{\\"tr\\":[{\\"i\\":0,\\"u\\":5},{\\"i\\":1,\\"u\\":15},{\\"i\\":2,\\"u\\":25},{\\"i\\":3,\\"u\\":35},{\\"i\\":4,\\"u\\":45},{\\"i\\":5,\\"u\\":55}],\\"dir\\":[[{\\"i\\":0,\\"u\\":5},0],[{\\"i\\":1,\\"u\\":15},1],[{\\"i\\":2,\\"u\\":25},2],[{\\"i\\":3,\\"u\\":35},3],[{\\"i\\":4,\\"u\\":45},4],[{\\"i\\":5,\\"u\\":55},5]],\\"set\\":[0,1,2,3,4,5]}",
  ],
  Array [
    "STEP",
    6,
    6,
    "{\\"tr\\":[{\\"i\\":0,\\"u\\":6},{\\"i\\":1,\\"u\\":16},{\\"i\\":2,\\"u\\":26},{\\"i\\":3,\\"u\\":36},{\\"i\\":4,\\"u\\":46},{\\"i\\":5,\\"u\\":56},{\\"i\\":6,\\"u\\":66}],\\"dir\\":[[{\\"i\\":0,\\"u\\":6},0],[{\\"i\\":1,\\"u\\":16},1],[{\\"i\\":2,\\"u\\":26},2],[{\\"i\\":3,\\"u\\":36},3],[{\\"i\\":4,\\"u\\":46},4],[{\\"i\\":5,\\"u\\":56},5],[{\\"i\\":6,\\"u\\":66},6]],\\"set\\":[0,1,2,3,4,5,6]}",
  ],
  Array [
    "UNDO",
  ],
  Array [
    "STEP",
    7,
    3,
    "{\\"tr\\":[{\\"i\\":0,\\"u\\":3},{\\"i\\":1,\\"u\\":13},{\\"i\\":2,\\"u\\":23},{\\"i\\":3,\\"u\\":33},{\\"i\\":4,\\"u\\":43},{\\"i\\":5,\\"u\\":53},{\\"i\\":6,\\"u\\":63},{\\"i\\":3,\\"u\\":33}],\\"dir\\":[[{\\"i\\":0,\\"u\\":3},0],[{\\"i\\":1,\\"u\\":13},1],[{\\"i\\":2,\\"u\\":23},2],[{\\"i\\":3,\\"u\\":33},3],[{\\"i\\":4,\\"u\\":43},4],[{\\"i\\":5,\\"u\\":53},5],[{\\"i\\":6,\\"u\\":63},6],[{\\"i\\":3,\\"u\\":33},3]],\\"set\\":[0,1,2,3,4,5,6]}",
  ],
  Array [
    "STEP",
    8,
    4,
    "{\\"tr\\":[{\\"i\\":0,\\"u\\":4},{\\"i\\":1,\\"u\\":14},{\\"i\\":2,\\"u\\":24},{\\"i\\":3,\\"u\\":34},{\\"i\\":4,\\"u\\":44},{\\"i\\":5,\\"u\\":54},{\\"i\\":6,\\"u\\":64},{\\"i\\":3,\\"u\\":34},{\\"i\\":4,\\"u\\":44}],\\"dir\\":[[{\\"i\\":0,\\"u\\":4},0],[{\\"i\\":1,\\"u\\":14},1],[{\\"i\\":2,\\"u\\":24},2],[{\\"i\\":3,\\"u\\":34},3],[{\\"i\\":4,\\"u\\":44},4],[{\\"i\\":5,\\"u\\":54},5],[{\\"i\\":6,\\"u\\":64},6],[{\\"i\\":3,\\"u\\":34},3],[{\\"i\\":4,\\"u\\":44},4]],\\"set\\":[0,1,2,3,4,5,6]}",
  ],
  Array [
    "SAVE2",
  ],
  Array [
    "STEP",
    9,
    5,
    "{\\"tr\\":[{\\"i\\":0,\\"u\\":5},{\\"i\\":1,\\"u\\":15},{\\"i\\":2,\\"u\\":25},{\\"i\\":3,\\"u\\":35},{\\"i\\":4,\\"u\\":45},{\\"i\\":5,\\"u\\":55},{\\"i\\":6,\\"u\\":65},{\\"i\\":3,\\"u\\":35},{\\"i\\":4,\\"u\\":45},{\\"i\\":5,\\"u\\":55}],\\"dir\\":[[{\\"i\\":0,\\"u\\":5},0],[{\\"i\\":1,\\"u\\":15},1],[{\\"i\\":2,\\"u\\":25},2],[{\\"i\\":3,\\"u\\":35},3],[{\\"i\\":4,\\"u\\":45},4],[{\\"i\\":5,\\"u\\":55},5],[{\\"i\\":6,\\"u\\":65},6],[{\\"i\\":3,\\"u\\":35},3],[{\\"i\\":4,\\"u\\":45},4],[{\\"i\\":5,\\"u\\":55},5]],\\"set\\":[0,1,2,3,4,5,6]}",
  ],
  Array [
    "STEP",
    10,
    6,
    "{\\"tr\\":[{\\"i\\":0,\\"u\\":6},{\\"i\\":1,\\"u\\":16},{\\"i\\":2,\\"u\\":26},{\\"i\\":3,\\"u\\":36},{\\"i\\":4,\\"u\\":46},{\\"i\\":5,\\"u\\":56},{\\"i\\":6,\\"u\\":66},{\\"i\\":3,\\"u\\":36},{\\"i\\":4,\\"u\\":46},{\\"i\\":5,\\"u\\":56},{\\"i\\":6,\\"u\\":66}],\\"dir\\":[[{\\"i\\":0,\\"u\\":6},0],[{\\"i\\":1,\\"u\\":16},1],[{\\"i\\":2,\\"u\\":26},2],[{\\"i\\":3,\\"u\\":36},3],[{\\"i\\":4,\\"u\\":46},4],[{\\"i\\":5,\\"u\\":56},5],[{\\"i\\":6,\\"u\\":66},6],[{\\"i\\":3,\\"u\\":36},3],[{\\"i\\":4,\\"u\\":46},4],[{\\"i\\":5,\\"u\\":56},5],[{\\"i\\":6,\\"u\\":66},6]],\\"set\\":[0,1,2,3,4,5,6]}",
  ],
  Array [
    "UNDO SAVE2",
  ],
  Array [
    "STEP",
    11,
    3,
    "{\\"tr\\":[{\\"i\\":0,\\"u\\":3},{\\"i\\":1,\\"u\\":13},{\\"i\\":2,\\"u\\":23},{\\"i\\":3,\\"u\\":33},{\\"i\\":4,\\"u\\":43},{\\"i\\":5,\\"u\\":53},{\\"i\\":6,\\"u\\":63},{\\"i\\":3,\\"u\\":33},{\\"i\\":4,\\"u\\":43},{\\"i\\":5,\\"u\\":53},{\\"i\\":6,\\"u\\":63},{\\"i\\":3,\\"u\\":33}],\\"dir\\":[[{\\"i\\":0,\\"u\\":3},0],[{\\"i\\":1,\\"u\\":13},1],[{\\"i\\":2,\\"u\\":23},2],[{\\"i\\":3,\\"u\\":33},3],[{\\"i\\":4,\\"u\\":43},4],[{\\"i\\":5,\\"u\\":53},5],[{\\"i\\":6,\\"u\\":63},6],[{\\"i\\":3,\\"u\\":33},3],[{\\"i\\":4,\\"u\\":43},4],[{\\"i\\":5,\\"u\\":53},5],[{\\"i\\":6,\\"u\\":63},6],[{\\"i\\":3,\\"u\\":33},3]],\\"set\\":[0,1,2,3,4,5,6]}",
  ],
  Array [
    "STEP",
    12,
    4,
    "{\\"tr\\":[{\\"i\\":0,\\"u\\":4},{\\"i\\":1,\\"u\\":14},{\\"i\\":2,\\"u\\":24},{\\"i\\":3,\\"u\\":34},{\\"i\\":4,\\"u\\":44},{\\"i\\":5,\\"u\\":54},{\\"i\\":6,\\"u\\":64},{\\"i\\":3,\\"u\\":34},{\\"i\\":4,\\"u\\":44},{\\"i\\":5,\\"u\\":54},{\\"i\\":6,\\"u\\":64},{\\"i\\":3,\\"u\\":34},{\\"i\\":4,\\"u\\":44}],\\"dir\\":[[{\\"i\\":0,\\"u\\":4},0],[{\\"i\\":1,\\"u\\":14},1],[{\\"i\\":2,\\"u\\":24},2],[{\\"i\\":3,\\"u\\":34},3],[{\\"i\\":4,\\"u\\":44},4],[{\\"i\\":5,\\"u\\":54},5],[{\\"i\\":6,\\"u\\":64},6],[{\\"i\\":3,\\"u\\":34},3],[{\\"i\\":4,\\"u\\":44},4],[{\\"i\\":5,\\"u\\":54},5],[{\\"i\\":6,\\"u\\":64},6],[{\\"i\\":3,\\"u\\":34},3],[{\\"i\\":4,\\"u\\":44},4]],\\"set\\":[0,1,2,3,4,5,6]}",
  ],
  Array [
    "UNDO SAVE",
  ],
  Array [
    5,
  ],
  Array [
    "UNDO SAVE1",
  ],
  Array [
    3,
  ],
]
`;
